lib/Parse/ParseExpr.cpp

```diff
@@ -3064,34 +3066,38 @@ ExprResult Parser::ParseBoundsCastExpression(IdentifierInfo &Ident,
             ParseRelativeBoundsClause(isError, Ident, Tok.getLocation());
         if (isError)
           SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
-	PT.consumeClose();
+        PT.consumeClose();
       }
     } else
       SkipUntil(tok::greater, StopAtSemi | StopBeforeMatch);
   }
 
-  SourceLocation EndLoc = Tok.getLocation();
+  SourceLocation RAngleBracketLoc = Tok.getLocation();
   if (ExpectAndConsume(tok::greater))
-    return ExprError(Diag(EndLoc, diag::note_matching) << tok::less);
+    return ExprError(Diag(RAngleBracketLoc, diag::note_matching) << tok::less);
 
   SourceLocation LParenLoc, RParenLoc;
   BalancedDelimiterTracker T(*this, tok::l_paren);
 
-  if (T.expectAndConsume(diag::err_expected_lparen_after, Ident.getNameStart()))
+  if (T.expectAndConsume(diag::err_expected_lparen_after, CastName))
     return ExprError();
+
   LParenLoc = T.getOpenLocation();
 
   // Parsing e1, e2, e3
   ExprResult E1(true), E2(true), E3(true);
 
-  E1 = ParseCastExpression(true);
-  if (E1.isInvalid()) {
+  E1 = ParseAssignmentExpression();
+  if (!E1.isInvalid())
+    E1 = Actions.CorrectDelayedTyposInExpr(E1.get());
+  else {
     SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
     return ExprError();
   }
 
   if (Tok.is(tok::comma)) {
     ConsumeToken();
+    syntax = BoundsCastExpr::SyntaxType::Count;
```

Comment:
I will make sure to stop parsing and try skip tokens and return ExprError() if there is an error about E1, E2 and E3.